_source/_posts/2022-08-31-state-of-java-project-loom.md

```diff
@@ -0,0 +1,263 @@
+# 'newCachedThreadPool' with 'defaultThreadFactory'
+0:11.52 real,   13.21 s user,   4.91 s sys,     157% 6019pu,    0 amem,         2215972 mmem
+```
+
+Not so bad. Now, let's do the same using virtual threads.
+
+```java
+try (var executor = Executors.newVirtualThreadPerTaskExecutor()) {
+    IntStream.range(0, 100_000).forEach(i -> executor.submit(() -> {
+        Thread.sleep(Duration.ofSeconds(1));
+        System.out.println(i);
+        return i;
+    }));
+}
+```
+
+If I run and time it, I get the below numbers.
+
+```bash
+0:02.62 real,   6.83 s user,    1.46 s sys,     316% 14840pu,   0 amem,         350268 mmem
+```
+
+This is far more performant than using platform threads with thread pools. Of course, these are simple use cases; both thread pools and virtual thread implementations can be further optimized for better performance, but that's not the point of this post.
+
+Running Java Microbenchmark Harness (JMH) with the same code gives the below results, and you can see that virtual threads outperform platform threads by a huge margin.
+
+```bash
+# Throughput
+Benchmark                             Mode  Cnt  Score   Error  Units
+LoomBenchmark.platformThreadPerTask  thrpt    5  0.362 ± 0.079  ops/s
+LoomBenchmark.platformThreadPool     thrpt    5  0.528 ± 0.067  ops/s
+LoomBenchmark.virtualThreadPerTask   thrpt    5  1.843 ± 0.093  ops/s
+
+# Average time
+Benchmark                             Mode  Cnt  Score   Error  Units
+LoomBenchmark.platformThreadPerTask   avgt    5  5.600 ± 0.768   s/op
+LoomBenchmark.platformThreadPool      avgt    5  3.887 ± 0.717   s/op
+LoomBenchmark.virtualThreadPerTask    avgt    5  1.098 ± 0.020   s/op
+```
+
+You can find the benchmark [source code on GitHub](https://github.com/deepu105/java-loom-benchmarks). Here are some other meaningful benchmarks for virtual threads:
+
+- An interesting benchmark using ApacheBench [on GitHub](https://github.com/ebarlas/project-loom-comparison) by [Elliot Barlas](https://twitter.com/ElliotBarlas)
+- A benchmark using Akka actors [on Medium](https://medium.com/@zakgof/a-simple-benchmark-for-jdk-project-looms-virtual-threads-4f43ef8aeb1) by [Alexander Zakusylo](https://medium.com/@zakgof)
+- JMH benchmarks for I/O and non-I/O tasks [on GitHub](https://github.com/colincachia/loom-benchmark) by [Colin Cachia](https://twitter.com/colincachia)
+
+### Structured Concurrency
+
+> Structured concurrency will be an incubator feature in Java 19.
+
+[Structured Concurrency](https://openjdk.org/jeps/428) aims to simplify multi-threaded and parallel programming. It treats multiple tasks running in different threads as a single unit of work, streamlining error handling and cancellation while improving reliability and observability. This helps to avoid issues like thread leaking and cancellation delays. Being an incubator feature, this might go through further changes during stabilization.
```

Comment:
```suggestion
[Structured concurrency](https://openjdk.org/jeps/428) aims to simplify multi-threaded and parallel programming. It treats multiple tasks running in different threads as a single unit of work, streamlining error handling and cancellation while improving reliability and observability. This helps to avoid issues like thread leaking and cancellation delays. Being an incubator feature, this might go through further changes during stabilization.
```
it's not a product name, so no need to uppercase